mypy/report.py

```diff
@@ -103,6 +104,115 @@ def on_finish(self) -> None:
 reporter_classes['linecount'] = LineCountReporter
 
 
+class LineCoverageVisitor(TraverserVisitor):
+    def __init__(self, source: List[str]) -> None:
+        self.source = source
+
+        # For each line of source, we maintain a pair of
+        #  * the indentation level of the surrounding function
+        #    (-1 if not inside a function), and
+        #  * whether the surrounding function is typed.
+        # Initially, everything is covered at indentation level -1.
+        self.lines_covered = [(-1, True) for l in source]
+
+    # The Python AST has position information for the starts of
+    # elements, but not for their ends. Fortunately the
+    # indentation-based syntax makes it pretty easy to find where a
+    # block ends without doing any real parsing.
+
+    # TODO: Handle line continuations (explicit and implicit) and
+    # multi-line string literals. (But at least line continuations
+    # are normally more indented than their surrounding block anyways,
+    # by PEP 8.)
+
+    def indentation_level(self, line_number: int) -> Optional[int]:
+        """Return the indentation of a line of the source (specified by
+        zero-indexed line number). Returns None for blank lines or comments."""
```

Comment:
This doesn't seem to handle certain idioms correctly. Example:

```
def f():
    x = (
1)  # actually not indented logically
    y = """
foo"""  # again not indented logically
    z = \
"bar"  # again not indented
```

Not sure how serious these issues are or if it's worth fixing them now, but at least the triple-quoted string case is pretty frequent in some code I work on (though I dislike this "idiom"). We could perhaps use a Python lexer to get more reliable indent level information.